test/spmd/tensor/test_api.py

```diff
@@ -0,0 +1,175 @@
+
+        module_to_replicate = MyModel(20, 20, device=self.device_type)
+        replica_module = distribute_module(
+            module_to_replicate, device_mesh, replicate_fn
+        )
+        for param in replica_module.parameters():
+            self.assertIsInstance(param, DTensor)
+            self.assertEqual(param.placements, replica_spec)
+
+        # only shard part of module, and rest of module should be replicate
+        def shard_fn(name, module):
+            if isinstance(module, nn.Linear) and (
+                name == "seq.0" or name == "seq.8"
+            ):
+                for name, param in module.named_parameters():
+                    dist_param = torch.nn.Parameter(
+                        distribute_tensor(param, device_mesh, shard_spec)
+                    )
+                    module.register_parameter(name, dist_param)
+
+        module_to_distribute = MyModel(20, 20, device=self.device_type)
+        dist_module = distribute_module(
+            module_to_distribute, device_mesh, shard_fn
+        )
+        for name, param in dist_module.named_parameters():
+            self.assertIsInstance(param, DTensor)
+            if name.startswith("seq.0") or name.startswith("seq.8"):
+                self.assertEqual(param.placements, shard_spec)
+            else:
+                self.assertEqual(param.placements, replica_spec)
+
+    @with_comms
+    def test_distribute_module_input_fn_output_fn(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+
+        # fully replicate all linear modules
+        module_to_replicate = MyModel(20, 1, device=self.device_type)
+
+        # mark input sharding on dim 0
+        def input_fn(inputs):
+            return DTensor.from_local(inputs[0], device_mesh, [Shard(0)])
+
+        def output_fn(outputs):
+            assert isinstance(outputs, DTensor)
+            return outputs.to_local()
+
+        replica_module = distribute_module(
+            module_to_replicate,
+            device_mesh,
+            input_fn=input_fn,
+            output_fn=output_fn,
+        )
+
+        input_tensor = torch.randn(5, 20, device=self.device_type)
+        local_out = replica_module(input_tensor)
+        self.assertIsInstance(local_out, torch.Tensor)
+        self.assertNotIsInstance(local_out, DTensor)
+
+        # full replicate (even on inputs)
+        model = MyModel(10, 10, device=self.device_type)
+
+        def replicate_input_fn(inputs):
+            return DTensor.from_local(inputs[0], device_mesh, [Replicate()])
+
+        replica_model = distribute_module(
+            model,
+            device_mesh,
+            input_fn=replicate_input_fn,
+        )
+        input = torch.randn(10, 10, requires_grad=True)
+        output = replica_model(input)
+        output.sum().backward()
+        param_grad = list(replica_model.parameters())[0].grad
+        self.assertTrue(isinstance(param_grad, DTensor))
```

Comment:
Should we add a test with a loss function and optimizer step as well?

Reply:
This is actually already been added in `test_megatron_example.py`